8347830: [premain] UseCompatibleCompressedOops is broken after merging with mainline

vnkozlov opened the pull request:
The issue was that we did not allocate space "noaccess prefix" before heap for read-only page when we setup type of encoding  with `UseCompatibleCompressedOops` flag. I update code which reserves heap for compressed oops to take into account this flag.

An other issue was that `UseCompatibleCompressedOops` flag was set during CDS arguments consistency checkup. But `UseCompressedOops` flag is set by GC ergonomics. I moved code to CDS ergonomics setting method.

During testing I hit assert that AOT code address table is missing some code blob address. But it was actually card table base address which pointed inside CodeCache. I added special case to check for it early.

Tested premain-tier1.

vnkozlov commented:
@ashu-mehra and @adinn please look.

ashu-mehra commented on src/hotspot/share/cds/cdsConfig.cpp:
```diff
@@ -120,6 +120,26 @@ void CDSConfig::ergo_initialize() {
+  if (UseCompressedOops && AOTCodeCache::is_caching_enabled()) {
+    FLAG_SET_ERGO_IF_DEFAULT(UseCompatibleCompressedOops, true);
+  } else if (!FLAG_IS_DEFAULT(UseCompatibleCompressedOops)) {
+    FLAG_SET_ERGO(UseCompatibleCompressedOops, false);
```
Why do we need to set it to false explicitly here? If it is not default shouldn't we use whatever value the user specified?

vnkozlov replied:
I consider the flag make sense only for AOT code and not for general usage. See description: [globals.hpp#L132](https://github.com/openjdk/leyden/blob/premain/src/hotspot/share/runtime/globals.hpp#L132)

We don't want generate more complex encoding if we don't generate/use AOT code. Or compressed oops are off.
We had also recently discussed how we should disable some AOT features if AOT caching is off - the conclusion was that we should silently ignore them (may be produce UL output on demand). That is why I used ERGO setting.

ashu-mehra replied:
okay

ashu-mehra commented on src/hotspot/share/memory/memoryReserver.cpp:
```diff
@@ -549,8 +549,17 @@ ReservedHeapSpace HeapReserver::Instance::reserve_compressed_oops_heap(const siz
+
+  bool unscaled  = false;
+  bool zerobased = false;
+  bool heapbased = UseCompatibleCompressedOops;
```
Use of `heapbased` is a bit confusing to me. Can we restructure it like this:
```
bool unscaled = false;
bool zerobased = false;
if (UseCompatibleCompressedOops) {
  unscaled = (heap_end_address <= (char*)UnscaledOopHeapMax);
  zerobased = (heap_end_address <= (char*)OopEncodingHeapMax);
}
size_t noaccess_prefix = (UseCompatibleCompressedOops || !zerobased) ? noaccess_prefix_size : 0;
```

vnkozlov replied:
I can do that. I got names from [CompressedOops::Mode](https://github.com/openjdk/leyden/blob/premain/src/hotspot/share/oops/compressedOops.hpp#L67) values

vnkozlov commented:
@ashu-mehra I refactored code as you asked.

vnkozlov commented:
@ashu-mehra I refactored code as you asked.

I looked on our old code where we added this flag and found that I missed 2 other places to check which I added now.

ashu-mehra reviewed (approved):
looks good to me

vnkozlov commented:
Thank you. @ashu-mehra

adinn reviewed (approved):
Looks good.

vnkozlov commented:
Thank you, @adinn

vnkozlov commented:
/integrate
